### PR TITLE
No kill exception

### PR DESCRIPTION
**Pull request by jtulach**
This change is taking the work done by Martin @Entlicher in #137 one step further. It hides the `QuitException` and `KillException` from the API and "piggy-backs" on the re-throw semantics of the `ThreadDeath`.

The removed `QuitException` and `KillException` were introduced after 0.11 version and thus their removal isn't (that) incompatible. Better to do it now then play the deprecations tricks once 0.12 is out.

@mlvdv by merging in this change you'd also accept Martin's changes from #137.

### CODE REVIEW

File: truffle/com.oracle.truffle.api.debug/src/com/oracle/truffle/api/debug/Debugger.java [hunk 1]
```diff
@@ -1059,10 +1060,14 @@ void executionEnded() {
...
-        } else {
-            return ACCESSOR.evalInContext(engine, ev, code, frameInstance.getCallNode(), frameInstance.getFrame(FrameAccess.MATERIALIZE, true).materialize());
+        try {
+            if (frameInstance == null) {
+                return ACCESSOR.evalInContext(engine, ev, code, debugContext.haltedNode, debugContext.haltedFrame);
+            } else {
+                return ACCESSOR.evalInContext(engine, ev, code, frameInstance.getCallNode(), frameInstance.getFrame(FrameAccess.MATERIALIZE, true).materialize());
+            }
+        } catch (KillException kex) {
+            throw new IOException("Killed");
```
Comment (chumer) [inline truffle/com.oracle.truffle.api.debug/src/com/oracle/truffle/api/debug/Debugger.java]: Please provide a better message. Something like "Evaluation was killed." Also initialize the cause in the IOException constructor.

**Comment by chumer** [comment]
Besides the one comment. Looks good.

File: truffle/com.oracle.truffle.api.debug/src/com/oracle/truffle/api/debug/Debugger.java [hunk 2]
```diff
@@ -1067,7 +1067,7 @@ Object evalInContext(SuspendedEvent ev, String code, FrameInstance frameInstance
                 return ACCESSOR.evalInContext(engine, ev, code, frameInstance.getCallNode(), frameInstance.getFrame(FrameAccess.MATERIALIZE, true).materialize());
             }
         } catch (KillException kex) {
-            throw new IOException("Killed");
+            throw (IOException) new IOException("Evaluation was killed.").initCause(kex);
```
Comment (chumer) [inline truffle/com.oracle.truffle.api.debug/src/com/oracle/truffle/api/debug/Debugger.java]: Why not just `new IOException("Evaluation was killed.", kex)`?

**Comment by jtulach** [comment]
My IDE even shows me a hint to switch to IOException constructor, but I haven't noticed it for some reason.

Martin @Entlicher is OK with the change. @chumer as well. @mlvdv has been busy so far. Time to merge tomorrow?

File: truffle/com.oracle.truffle.tools.debug.shell/src/com/oracle/truffle/tools/debug/shell/client/REPLRemoteCommand.java [hunk 3]
```diff
@@ -579,6 +579,7 @@ void processReply(REPLClientContext context, REPLMessage[] replies) {
             } else {
                 context.displayFailReply(replies[0].get(REPLMessage.DISPLAY_MSG));
             }
+            throw new REPLContinueException();
```
Comment (woess) [inline truffle/com.oracle.truffle.tools.debug.shell/src/com/oracle/truffle/tools/debug/shell/client/REPLRemoteCommand.java]: I guess this was a bug in the REPL?

Comment (jtulach) [inline truffle/com.oracle.truffle.tools.debug.shell/src/com/oracle/truffle/tools/debug/shell/client/REPLRemoteCommand.java]: I don't know. This change is from original Martin @entlicher commit - which was part of #137 pull request.

Comment (woess) [inline truffle/com.oracle.truffle.tools.debug.shell/src/com/oracle/truffle/tools/debug/shell/client/REPLRemoteCommand.java]: Now that I think about it, it's probably due to the kill -> prepareKill change.

Comment (entlicher) [inline truffle/com.oracle.truffle.tools.debug.shell/src/com/oracle/truffle/tools/debug/shell/client/REPLRemoteCommand.java]: Exactly.  It wasn't a bug. The code was interrupted by KillException before. After we prepare for the kill only, we need to continue to actually finish the kill.

Comment (woess) [inline truffle/com.oracle.truffle.tools.debug.shell/src/com/oracle/truffle/tools/debug/shell/client/REPLRemoteCommand.java]: That's what I thought. Thanks for the clarification.

**Pull request merged**